src/core/decoder/decoder.ml

```diff
@@ -419,65 +419,94 @@ let get_stream_decoder ~ctype mime =
 (** {1 Helpers for defining decoders} *)
 
 let mk_buffer ~ctype generator =
-  let has_audio = Frame.Fields.mem Frame.Fields.audio ctype in
-  let has_video = Frame.Fields.mem Frame.Fields.video ctype in
-
-  let put_pcm =
-    if has_audio then (
-      let resampler = Decoder_utils.samplerate_converter () in
-      let current_channel_converter = ref None in
-      let current_dst = ref None in
-
-      let mk_channel_converter dst =
-        let c = Decoder_utils.channels_converter dst in
-        current_dst := Some dst;
-        current_channel_converter := Some c;
-        c
+  let audio_handlers = Hashtbl.create 0 in
+  let video_handlers = Hashtbl.create 0 in
+
+  let get_audio_handler ~field =
+    try Hashtbl.find audio_handlers field
+    with Not_found ->
+      let handler =
+        if Frame.Fields.mem field ctype then (
+          let resampler = Decoder_utils.samplerate_converter () in
+          let current_channel_converter = ref None in
+          let current_dst = ref None in
+
+          let mk_channel_converter dst =
+            let c = Decoder_utils.channels_converter dst in
+            current_dst := Some dst;
+            current_channel_converter := Some c;
+            c
+          in
+
+          let get_channel_converter () =
+            let dst =
+              channel_layout (Frame.Fields.find Frame.Fields.audio ctype)
+            in
+            match !current_channel_converter with
+              | None -> mk_channel_converter dst
+              | Some _ when !current_dst <> Some dst -> mk_channel_converter dst
+              | Some c -> c
+          in
+          fun ~samplerate data ->
+            let data, _, _ = resampler ~samplerate data 0 (Audio.length data) in
+            let data = (get_channel_converter ()) data in
+            let data = Content.Audio.lift_data data in
+            (*
+Printf.printf "put pcm for %s\n%!" (Frame.Fields.string_of_field field);
+Printf.printf "Buffer content_type: %s\n%!" (Frame.string_of_content_type (Generator.content_type generator));
+Printf.printf "Buffer len: %d\n%!" (Generator.length generator);
+Printf.printf "Buffer buffered len: %d\n%!" (Generator.buffered_length generator);
+Frame.Fields.iter (fun field c -> Printf.printf "%s len: %d\n%!" (Frame.Fields.string_of_field field) (Content.length c)) (Generator.peek generator);
+*)
```

Comment:
Remove this debug, unless you really think that it could be useful in the future:
```suggestion
```